`error: redeclaration of ‘qX60gensym0_’ with no linkage` with nested templates

### Description

```nim
template c(w: int): int =
  let q = w
  0

template t(r: (int, int); x: int) =
  for _ in r.fields:
    let w = x

proc k() =
  t((0, 0), c(0))

k()
```

Applies across Cartesian product of
- {`v2.0.10`, `version-2-0`, `version-2-2`, `devel`} x 
- {`nim c`, `nim cpp`} x
- {`--mm:refc`, `--mm:orc`}


### Nim Version

```
Nim Compiler Version 2.0.10 [Linux: amd64]
Compiled at 2024-10-21
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: e941ee15be775fe3c46db1bed9b4f41c7dfb1334
active boot switches: -d:release
```

```
Nim Compiler Version 2.0.11 [Linux: amd64]
Compiled at 2024-10-21
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: e334aad20220ddb5fad63002102d5557464f48c1
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.0 [Linux: amd64]
Compiled at 2024-10-21
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 78983f1876726a49c69d65629ab433ea1310ece1
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.1 [Linux: amd64]
Compiled at 2024-10-21
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 67442471ae4504bc074c8bb02308613fadc199f8
active boot switches: -d:release
```

### Current Output

```text
/tmp/tmp.Yj3mj2C0HM/@mn.nim.c: In function ‘k__n_u8’:
/tmp/tmp.Yj3mj2C0HM/@mn.nim.c:67:12: error: redeclaration of ‘qX60gensym0_’ with no linkage
   67 |         NI qX60gensym0_;
      |            ^~~~~~~~~~~~
/tmp/tmp.Yj3mj2C0HM/@mn.nim.c:65:12: note: previous declaration of ‘qX60gensym0_’ with type ‘NI’ {aka ‘long int’}
   65 |         NI qX60gensym0_;
      |            ^~~~~~~~~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread   -I/tmp/nim/lib -I/tmp -o /tmp/tmp.Yj3mj2C0HM/@mn.nim.c.o /tmp/tmp.Yj3mj2C0HM/@mn.nim.c' failed with exit code: 1
```


### Expected Output

```text
No invalid C codegen
```


### Known Workarounds

_No response_

### Additional Information

_No response_

--- a/compiler/semfields.nim
+++ b/compiler/semfields.nim
@@ -18,6 +18,12 @@ type
     replaceByFieldName: bool
     c: PContext
 
+proc wrapNewScope(c: PContext, n: PNode): PNode {.inline.} =
+  # use `if true` to not interfere with `break`
+  let trueLit = newIntLit(c.graph, n.info, 1)
+  trueLit.typ() = getSysType(c.graph, n.info, tyBool)
+  result = newTreeI(nkIfStmt, n.info, newTreeI(nkElifBranch, n.info, trueLit, n))
+
 proc instFieldLoopBody(c: TFieldInstCtx, n: PNode, forLoop: PNode): PNode =
   if c.field != nil and isEmptyType(c.field.typ):
     result = newNode(nkEmpty)
@@ -72,7 +78,9 @@ proc semForObjectFields(c: TFieldsCtx, typ, forLoop, father: PNode) =
     )
     openScope(c.c)
     inc c.c.inUnrolledContext
-    let body = instFieldLoopBody(fc, lastSon(forLoop), forLoop)
+    var body = instFieldLoopBody(fc, lastSon(forLoop), forLoop)
+    # new scope for each field
+    body = wrapNewScope(c.c, body)
     father.add(semStmt(c.c, body, {}))
     dec c.c.inUnrolledContext
     closeScope(c.c)
@@ -148,6 +156,8 @@ proc semForFields(c: PContext, n: PNode, m: TMagic): PNode =
           replaceByFieldName: m == mFieldPairs
       )
       var body = instFieldLoopBody(fc, loopBody, n)
+      # new scope for each field
+      body = wrapNewScope(c, body)
       inc c.inUnrolledContext
       stmts.add(semStmt(c, body, {}))
       dec c.inUnrolledContext

--- a/compiler/semfields.nim
+++ b/compiler/semfields.nim
@@ -20,6 +20,9 @@ type
 
 proc wrapNewScope(c: PContext, n: PNode): PNode {.inline.} =
   # use `if true` to not interfere with `break`
+  # just opening scope via `openScope(c)` isn't enough,
+  # a scope has to be opened in the codegen as well for reused
+  # template instantiations
   let trueLit = newIntLit(c.graph, n.info, 1)
   trueLit.typ() = getSysType(c.graph, n.info, tyBool)
   result = newTreeI(nkIfStmt, n.info, newTreeI(nkElifBranch, n.info, trueLit, n))
@@ -79,7 +82,7 @@ proc semForObjectFields(c: TFieldsCtx, typ, forLoop, father: PNode) =
     openScope(c.c)
     inc c.c.inUnrolledContext
     var body = instFieldLoopBody(fc, lastSon(forLoop), forLoop)
-    # new scope for each field
+    # new scope for each field that codegen should know about:
     body = wrapNewScope(c.c, body)
     father.add(semStmt(c.c, body, {}))
     dec c.c.inUnrolledContext
@@ -156,7 +159,7 @@ proc semForFields(c: PContext, n: PNode, m: TMagic): PNode =
           replaceByFieldName: m == mFieldPairs
       )
       var body = instFieldLoopBody(fc, loopBody, n)
-      # new scope for each field
+      # new scope for each field that codegen should know about:
       body = wrapNewScope(c, body)
       inc c.inUnrolledContext
       stmts.add(semStmt(c, body, {}))

--- a/tests/system/tfields.nim
+++ b/tests/system/tfields.nim
@@ -106,3 +106,24 @@ block timplicit_with_partial:
     echo x
 
   foo(FooTask())
+
+block:
+  var innerCount = 0
+  var outerCount = 0
+  template c(w: int): int =
+    let q = w
+    inc innerCount
+    0
+
+  template t(r: (int, int); x: int) =
+    for _ in r.fields:
+      let w = x
+      doAssert w == 0
+      dec outerCount
+
+  proc k() =
+    t((0, 0), c(0))
+
+  k()
+  doAssert innerCount == 2
+  doAssert outerCount == -2

--- a/tests/system/tfields.nim
+++ b/tests/system/tfields.nim
@@ -107,7 +107,7 @@ block timplicit_with_partial:
 
   foo(FooTask())
 
-block:
+block: # issue #24338
   var innerCount = 0
   var outerCount = 0
   template c(w: int): int =
@@ -123,6 +123,29 @@ block:
 
   proc k() =
     t((0, 0), c(0))
+
+  k()
+  doAssert innerCount == 2
+  doAssert outerCount == -2
+
+block: # issue #24338 with object
+  type Foo = object
+    x, y: int
+  var innerCount = 0
+  var outerCount = 0
+  template c(w: int): int =
+    let q = w
+    inc innerCount
+    0
+
+  template t(r: Foo; x: int) =
+    for _ in r.fields:
+      let w = x
+      doAssert w == 0
+      dec outerCount
+
+  proc k() =
+    t(Foo(x: 0, y: 0), c(0))
 
   k()
   doAssert innerCount == 2

